### PR TITLE
proc macro to generate memory-efficient gate

### PR DESCRIPTION
**Pull request by taikiy**
Procedural macro to implement the `Step` and `StepNarrow` traits and generate a memory-efficient gate.

The goal is to generate a state transition graph and the corresponding `StepNarrow` implementations for the IPA protocol. This macro assumes that a complete IPA steps file exists in the repo at the location specified by `STEPS_FILE`. The steps file can be generated by running `collect_steps.py`.

The steps file contains a list of narrowed steps, where each line represents a hierarchy of narrowed steps delimited by "/". For example, the following lines represent a hierarchy of narrowed steps ("=> x" parts are not in the file. They are added in here as comments to indicate their potential IDs when the macro parses the file):

```
    RootStep                                => 0
    RootStep/StepA::A1                      => 1
    RootStep/StepA::A1/StepB::B1            => 2
    RootStep/StepA::A1/StepB::B2            => 3
    RootStep/StepC::C1                      => 4
    RootStep/StepC::C1/StepD::D1            => 5
    RootStep/StepC::C1/StepD::D1/StepA::A2  => 6
    RootStep/StepC::C2                      => 7
```

From these lines, we want to generate `StepNarrow` implementations for each step.

```
    impl StepNarrow<StepA> for Compact {
        fn narrow(&self, step: &StepA) -> Self {
            Self(match (self.0, step.as_ref()) {
                (0, "A1") => 1,
                (5, "A2") => 6,
                _ => panic!("invalid state transition"),
            })
        }
    }
    impl StepNarrow<StepB> for Compact {
        fn narrow(&self, step: &StepB) -> Self {
            Self(match (self.0, step.as_ref()) {
                (1, "B1") => 2,
                (1, "B2") => 3,
                _ => panic!("invalid state transition"),
            })
        }
    }
    ...
```

Currently, this derive macro assumes it annotates the `Compact` struct defined in `src/protocol/step/compact.rs`. The `Compact` struct is a wrapper around a `u16` value that represents the current state of the IPA protocol.

In the future, we might change the macro to annotate each step in the IPA protocol. The macro will then generate both `Descriptive` and `Compact` implementations for the step. However, that kind of derive macro needs to be able to analyze the fully qualified module path of the step at runtime. This is because there are many locally-defined `Step` enums in IPA, and we need to disambiguate them. There are two ways to solve this. One is to use a feature to resolve the fully qualified module path of a Step enum/struct, but the feature is in nightly as of June 2023. Another way is to define globally unique step names.

### CODE REVIEW

**Comment by martinthomson** [comment]
One comment, just based on your commentary (not the patch, I am out of energy for the week).

```rust
    impl StepNarrow<StepA> for Compact {
        fn narrow(&self, step: &StepA) -> Self {
            Self(match (self.0, step.as_ref()) {
                (0, "A1") => 1,
```

This would be better as:

```rust
    impl StepNarrow<StepA> for Compact {
        fn narrow(&self, step: &StepA) -> Self {
            Self(match (self.0, step) {
```

And, to confirm.  Do you have annotations for enum arms that take integers, like this:

```rust
enum StepC {
  DoX,
  DoY(u32),
}
```

It would be OK if you only have code for `BitOpStep` and friends.

File: ipa-macros/tests/derive_gate/common/mod.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            StepA::ThisStateDoesNotExistInTheMap => "this-state-does-not-exist-in-the-map",
+        }
+    }
+}
+impl Step for StepB {}
+impl AsRef<str> for StepB {
+    fn as_ref(&self) -> &str {
+        match self {
+            StepB::Bar => "bar",
+            StepB::Baz => "baz",
```
Comment (martinthomson) [inline ipa-macros/tests/derive_gate/common/mod.rs:30]:
```suggestion
            Self::Bar => "bar",
            Self::Baz => "baz",
```

File: ipa-macros/src/derive_gate/mod.rs [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+//     a proc macro, but that requires a feature currently in nightly (mentioned in #1).
+//
+// 5. num-multi-bits
+//   - `num-multi-bits` also changes the state transition map. We could generate the steps file for all
+//     possible values of `num-multi-bits, but that will make the file huge. We can probably just stick to
+//     the current empirical best value of 3. We could also try to read the value from the source or the
+//     config file, and generate the steps file accordingly. However, the value could change after the steps
+//     file is generated, so we need to make sure that the steps file is always up to date somehow.
+//
+// 6. Root step
```
Comment (martinthomson) [inline ipa-macros/src/derive_gate/mod.rs:70]: We might need a separate type and proc macro for that.

Comment (taikiy) [reply ipa-macros/src/derive_gate/mod.rs:70]: This was more about whether we need to dynamically generate a string representation of root step(s), i.e., does query ID "X" must be generated by `Compact` gate's `AsRef<str>` implementation? If we do not require the root step in a real world scenario, we don't need to worry about this. `Descriptive` gate has `id` field which embeds `protocol/run-0` as the root step, so I was wondering if that was also needed for `Compact` gate.

File: ipa-macros/src/derive_gate/mod.rs [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+struct StepMetaData {
+    id: u16,
+    depth: u8,
+    module: String,
+    name: String,
+    path: String,
+}
+
+impl StepMetaData {
+    pub fn new(id: u16, depth: u8, module: &str, name: &str, path: &str) -> Self {
```
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/derive_gate/mod.rs:136]: In this particular file, we don't have "String -> &str", but I can foresee that happening. Modified as suggested.

File: ipa-macros/src/derive_gate/mod.rs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+            module: module.to_owned(),
+            name: name.to_owned(),
+            path: path.to_owned(),
+        }
+    }
+}
+
+const TARGET_CRATE: &str = "ipa";
+#[cfg(not(feature = "trybuild"))]
+const STEPS_FILE_NAME: &str = "steps.txt";
```
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/derive_gate/mod.rs:149]: `collect_steps.py` outputs the steps to stdout, so it's up to the user to redirect and name the file. It'd be great if we can detect that `Compact` code needs to be re-generated, we could automate running `collect_steps.py`, saving the steps file to the pre-defined path, and use the env var to replace `STEPS_FILE_NAME`.

File: ipa-macros/src/derive_gate/steps.test.txt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+StepA::foo
```
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/derive_gate/steps.test.txt:1]: I like the idea but couldn't think of what to add in the schematized file. Maybe one thing we can add now is `TARGET_CRATE`. Definitely worth thinking in the future when we try to solve #704 and adding other improvements.

File: ipa-macros/src/tree.rs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    pub fn get_parent(&self) -> Option<Node<T>> {
+        let my_parent_weak = self.parent.borrow();
+        if let Some(my_parent_ref) = my_parent_weak.upgrade() {
+            Some(Node {
+                data: my_parent_ref,
+            })
+        } else {
+            None
+        }
```
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/tree.rs:72]: The difference here is that `my_parent_ref` encloses `InnerNode` but the return type is `Node`. I could implement From<> but this is the only place we would use that.

File: ipa-macros/src/tree.rs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+            my_children.push(new_child.get_copy());
+        } // drop the borrow
+        {
+            let mut childs_parent = new_child.data.parent.borrow_mut();
+            *childs_parent = Rc::downgrade(&self.get_copy());
+        } // drop the borrow
+        new_child
+    }
+
+    pub fn get_children(&self) -> Vec<Node<T>> {
```
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/tree.rs:56]: Let me take this as a homework. Not sure how to return an iterator while keeping `borrow().iter()`.

File: ipa-macros/src/tree.rs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+#[derive(Clone, Debug)]
+pub struct InnerNode<T> {
+    value: T,
+    parent: Parent<T>,
+    children: Children<T>,
+}
+
+impl<T> Node<T> {
+    pub fn new(value: T) -> Node<T> {
```
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/tree.rs:28]: Done!

File: ipa-macros/src/tree.rs [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+};
+
+#[derive(Clone, Debug)]
+pub struct Node<T> {
+    data: InnerNodeRef<T>,
+}
+
+// We want to make sure that once a parent node is dropped, all its children are dropped as well.
+// To do that, we use `Rc<>` for children, and `Weak<>` for parent. (`Rc` because it's single-threaded)
+type InnerNodeRef<T> = Rc<InnerNode<T>>;
```
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/tree.rs:15]: Yes, it's a bit confusing...

The type of `Node::data::parent` is

`Rc<RefCell<Weak<InnerNode<T>>>>`

and `Node::data::children` is

`Rc<RefCell<Vec<Rc<InnerNode<T>>>>>`,

but `Node::data::value` is

`Rc<T>`.

We can keep the `value` immutable while able to mutate parent/child edges via `add_child`.

File: ipa-macros/src/tree.rs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+        Node {
+            data: Rc::new(new_node),
+        }
+    }
+
+    fn get_copy(&self) -> InnerNodeRef<T> {
+        Rc::clone(&self.data)
+    }
+
+    pub fn add_child(&self, value: T) -> Node<T> {
```
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/tree.rs:43]: Yes, that would make this part of the code a bit more readable. I think hiding all Rc/Weak stuff from the users of `Node` is a pretty nice feature though.

File: ipa-macros/src/tree.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (taikiy) [reply ipa-macros/src/tree.rs:72]: Fixed!

**Pull request merged**